### DISCUSSION TITLE
Very slow writing to file

### DISCUSSION THREAD

**Discussion by czarnikjak** [micropython/micropython#12442] [RP2040 / Pico] [answered]
I am not sure what i am doing wrong, but I am experiencing very slow writing to a file located on built-in flash of Pico.
The file is 250KB in size, and I am writing only 128 bytes to it (after seeking to a specific offset).

I measured the time it takes to write  with the ticks_us diff, and it takes over 3 seconds! Please tell me I am doing something wrong and this should be much faster...

```python

     with open('myfile', mode='r+b') as file:
          file.seek(position)
          t = time.ticks_us()
          file.write(myByteArray[128:256])
      
      print(time.ticks_diff(time.ticks_us(), t))
```

**Comment by jimmo** [comment]
@czarnikjak Is this easily reproducible? Does the slow write happen every time. Are you able to provide a simple script I can run here to test the same thing?

For some background... 3 seconds seems too long. Sometimes writes will be slower as the LittleFS filesystem has to rebalance the flash and do housekeeping, but this should not occur on every write.

**Reply by czarnikjak** [reply]
Hi @jimmo ,
Repeatable every time on my Pico W,
Below is a test script. File is 250KB in size.

```python

import time

myByteArray = bytearray(512)
position = 0
myFile = '1.file'

for i in range(20): 
  print('Position ',position)
  file = open(myFile, mode='r+b')
  file.seek(position)  
  file.write(myByteArray[128:256])
  t = time.ticks_us()
  file.flush()
  print(time.ticks_diff(time.ticks_us(), t))
  file.close()
  position = position + 10000

```
 
And here is the output, you can see writing 128bytes near the beginning of the file takes 3.5s, as you go further down the file its faster, about 1second if you writing 128bytes near the 200KB offset.

Position  0
3501963
Position  10000
3431295
Position  20000
3302977
Position  30000
3148699
Position  40000
3003530
Position  50000
2876461
Position  60000
2720220
Position  70000
2557618
Position  80000
2484944
Position  90000
2334595
Position  100000
2145765
Position  110000
2074833
Position  120000
1866241
Position  130000
1752181
Position  140000
1596832
Position  150000
1490915
Position  160000
1326186
Position  170000
1180414
Position  180000
1032127
Position  190000
926753

**Comment by robert-hh** [comment]
It seems to be a matter of the file system. With LFS I get the similar numbers on RP2 or e.g. MIMXRT. With a FAT file system the write times are consistent at about 140ms. It seems that with LFS the whole tail of the file is rewritten.

**Reply by czarnikjak** [reply]
Hmm...with LFS i get file system with size of about 860KB or so, but when trying to create FAT fs, my size seems to be only 89KB with the following parameters...

>>>os.umount('/')
>>> os.VfsFat.mkfs(flash)
>>> os.mount(flash, '/')
>>> print(os.statvfs('/'))
(512, 512, 178, 178, 178, 0, 0, 0, 0, 255)

There seems to be no options to customize anything, like a block size. Any ideas how to create larger FAT fs on Pico?

**Reply by robert-hh** [reply]
The standard block size for FAT is 4096. To get a FAT file system, try:
```
import os, rp2
bdev = rp2.Flash()
os.VfsFat.mkfs(bdev)
vfs = os.VfsFat(bdev)
os.mount(vfs, "/")
```
For frequent writes, FAT is bad, since it does not support wear leveling. The flash vendor usually guarantee ~100000 block erase cycles. In a test I tried to verify that. The flash chip broke after ~5 Million at best conditions.
If you need to write frequently to that file, better use another option, like either the one Jimmo suggested, of an external Micro-SD card. These support wear-leveling even with FAT and are generally way faster.

**Reply by czarnikjak** [reply]
Thanks, it defo creates 512 block size, with an overall FS size of only 89KB

>>> os.umount('/')
>>> bdev = rp2.Flash()
>>> os.VfsFat.mkfs(bdev)
>>> vfs = os.VfsFat(bdev)
>>> os.mount(vfs, "/")
>>> print(os.statvfs('/'))
(512, 512, 178, 178, 178, 0, 0, 0, 0, 255)

I will try splitting the file into little chunks like jimmo suggested.

**Reply by robert-hh** [reply]
Somehow the wring block size is reported. At a ARDUINO_RP2040_NANO_CONNECT the block size is reported as 4096, and bdev.ioctl() also reports 4096 as block size. maybe that's a bug to be fixed.
Anyhow, LFS is the better choice for flash based small file systems. When splitting files, you have to find the balance between the number of files and the file size. With many files, LFS slows down as well. But I just ran a quick test on a PICO W, creating, reading & deleting a series of files. With 100 files it was not a problem. Creating a short file filled with random numbers took ~74ms in average. With a FAT file system, it took ~260ms.

**Comment by jimmo** [comment, marked as answer]
FYI @czarnikjak @robert-hh this behaviour in LittleFS is described here https://github.com/littlefs-project/littlefs/issues/244

Suggested workaround is to break your data up into smaller files.

**Reply by jimmo** [reply]
Also thank you @czarnikjak for the very detailed report and repro!
